clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java

```diff
@@ -68,36 +74,117 @@ public Builder(String groupId,
+                        topic.partitionIndexes().add(topicPartition.partition());
+                        offsetFetchRequestTopicMap.put(topicName, topic);
+                    }
+                    topics = new ArrayList<>(offsetFetchRequestTopicMap.values());
                 } else {
-                    log.trace("Fallback the requireStable flag to false as broker " +
-                                  "only supports OffsetFetchRequest version {}. Need " +
-                                  "v7 or newer to enable this feature", version);
-
-                    return new OffsetFetchRequest(data.setRequireStable(false), version);
+                    topics = ALL_TOPIC_PARTITIONS_BATCH;
                 }
+                groups.add(new OffsetFetchRequestGroup()
+                    .setGroupId(entry.getKey())
+                    .setTopics(topics));
             }
+            this.data = new OffsetFetchRequestData()
+                .setGroupIds(groups)
+                .setRequireStable(requireStable);
+            this.throwOnFetchStableOffsetsUnsupported = throwOnFetchStableOffsetsUnsupported;
+        }
+
+        @Override
+        public OffsetFetchRequest build(short version) {
+            if (version < 8) {
+                if (data.groupIds().size() > 1) {
+                    throw new NoBatchedOffsetFetchRequestException("Broker does not support"
+                        + " batching groups for fetch offset request on version " + version);
+                }
+                OffsetFetchRequestData oldDataFormat = null;
+                if (!data.groupIds().isEmpty()) {
+                    OffsetFetchRequestGroup group = data.groupIds().get(0);
+                    String groupName = group.groupId();
+                    List<OffsetFetchRequestTopics> topics = group.topics();
+                    List<OffsetFetchRequestTopic> oldFormatTopics = null;
+                    if (topics != null) {
+                        oldFormatTopics = topics
+                            .stream()
+                            .map(t ->
+                                new OffsetFetchRequestTopic()
+                                    .setName(t.name())
+                                    .setPartitionIndexes(t.partitionIndexes()))
+                            .collect(Collectors.toList());
+                    }
+                    oldDataFormat = new OffsetFetchRequestData()
+                        .setGroupId(groupName)
+                        .setTopics(oldFormatTopics)
+                        .setRequireStable(data.requireStable());
+                }
+                if (isAllTopicPartitions() && version < 2) {
+                    throw new UnsupportedVersionException("The broker only supports OffsetFetchRequest " +
+                        "v" + version + ", but we need v2 or newer to request all topic partitions.");
+                }
+                if (data.requireStable() && version < 7) {
```

Comment:
Again better to move this to the top as well so that we have smaller sequence of if statements rather than nested ifs.